lib/spatial_features/importers/shapefile.rb

```diff
@@ -49,12 +50,29 @@ def data_from_wkt(wkt, proj4)
         SQL
       end
 
-
       def file
-        @file ||= begin
+        @file ||= (requested_shp_file || possible_shp_files.first)
+      end
+
+      # a zip archive may contain multiple SHP files
+      def possible_shp_files
+        @possible_shp_files ||= begin
           validate_file!
-          Download.open(archive, unzip: /\.shp$/, downcase: true)
+          Download.open_each(archive, unzip: /\.shp$/, downcase: true)
+        end
+      end
+
+      def requested_shp_file
+        return unless @shp_file_path
+        expected_match = "*/#{@shp_file_path}"
+        file = possible_shp_files.find do |candidate|
+          ::File.fnmatch?(expected_match, candidate.path)
         end
+
+        raise SpatialFeatures::Unzip::PathNotFound, "could not find #{@shp_file_path} in #{possible_shp_files.map(&:path)}" \
+          unless file
+
+        file
```

Comment:
Unless it's a one-liner, I'm a fan of explicit `return` statements.

Reply:
👌